docs/articles/persistence/custom-persistence-provider.md

```diff
@@ -0,0 +1,467 @@
+
+Journals should aim to persist events in-order for a given `persistenceId` as otherwise in case of a failure, the persistent state may end up being inconsistent.
+
+##### AtomicWrite
+
+Each `AtomicWrite` message contains a `Payload` property that contains:
+
+* a single `IPersistentRepresentation` that corresponds to the event that was passed to the `Eventsourced.Persist<TEvent>(TEvent, Action<TEvent>)` method of the `PersistentActor`, or
+* it contains several `IPersistentRepresentation` that correspond to the events that were passed to the `Eventsourced.PersistAll<TEvent>(IEnumerable<TEvent>, Action<TEvent>)` method of the `PersistentActor`.
+
+Note that while `AtomicWrite.Payload` is declared as an object property, the type of this object instance will always be an `ImmutableList<IPersistentRepresentation>`.
+
+All `Persistent` of the `AtomicWrite` must be written to the data store atomically, i.e. all or none must be stored.
+
+If the journal (data store) cannot support atomic writes of multiple events it should reject such writes with a `NotSupportedException` describing the issue. This limitation should also be documented by the journal plugin.
+
+##### Handling Failures
+
+If there are failures when storing any of the messages in the batch the returned `Task` must be completed with failure. The `Task` must only be completed with success when all messages in the batch have been confirmed to be stored successfully, i.e. they will be readable, and visible, in a subsequent replay. If there is uncertainty about if the messages were stored or not the `Task` must be completed with failure.
+
+Data store connection problems must be signaled by completing the `Task` with failure.
+
+##### Handling `AtomicWrite` Result
+
+The journal can also signal that it rejects individual messages (`AtomicWrite`) by the returned `Task`.
+
+* It is possible but not mandatory to reduce the number of allocations by returning null for the happy path, i.e. when no messages are rejected.
+* Otherwise the returned list must have as many elements as the input `messages`.
+* Each result element signals if the corresponding `AtomicWrite` is rejected or not, with an exception describing the problem.
+* Rejecting a message means it was not stored, i.e. it must not be included in a later replay.
+* Rejecting a message is typically done before attempting to store it, e.g. because of serialization error.
+* Data store connection problems must not be signaled as rejections.
+
+##### Code Sample
+
+`InsertEventSql` in this example refers to this SQL query statement:
+
+[!code-csharp[InsertEventSql](../../../src/examples/Akka.Persistence.Custom/Journal/SqliteJournal.cs?name=InsertEventSql "InsertEventSql SQL Statement")]
+
+SQLite code example:
+
+[!code-csharp[WriteMessagesAsync](../../../src/examples/Akka.Persistence.Custom/Journal/SqliteJournal.cs?name=WriteMessagesAsync "WriteMessagesAsync method implementation")]
+
+#### DeleteMessagesToAsync
+
+```c#
```

Comment:
Do we leave this as it is or make use of the range selection though the line number may change in the future?

Reply:
see above